### DISCUSSION TITLE
Is there a guide for a beginner developer?

### DISCUSSION THREAD

**Discussion by iiicakeeiii** [musescore/MuseScore#19314] [Q&A] [answered]
Hi,

   I'm new to programming and wanted to have a full program to look at as I learn. Seeing musescore on here peaked my interest a bit. Is there a guide to help me look around?

**Comment by MarcSabatella** [comment]
Do you mean a guide to programming?  Or a guide to using Musescore?  For the former, see the info under Contribute / Development on musescore.org, although it's meant more for experienced programmers who just need to learn the specifics of how to contribute to the MsueScore project than for teaching programming.

For help with MsueScore itself, see Help / Online handbook from within the program itself.

**Reply by iiicakeeiii** [reply]
Sorry I meant the former, but understood I'll take a look at the development page on the site. 

So what experience level should someone be at (beginner-intermediate? intermediate-advanced?) before they could actually follow/contribute to the project? 

Or is it more of like I should just pop my head in one of the repositories now and then to see if I get a grasp of what's happening.
Thanks for the reply,

**Reply by cbjeukendrup** [reply, marked as answer]
It's not like you'd need a certain skill level in order to contribute (as far as skill levels can even be defined...); there are a lot of different possible tasks, ranging from trivial to very complicated. 

I can say though that it's useful to have some basic knowledge of programming, and some knowledge of C++ specifically, so that you can at least _read_ it. 
Once you have that, you're ready to start learning and contributing. Here are some tips:
- just look at some existing code and try to see if you can understand it
- if there's something you don't understand yet, look it up in the case that it's something from C++ or from Qt, or ask us if it's something specific to MuseScore
- pick one of the issues from https://github.com/orgs/musescore/projects/30, and try if you can find the relevant code; try to write some new code based on what you've seen. That's a very effective way to learn, in my experience!

Also, for how to get the code compiling and running, this information is crucial: https://github.com/musescore/MuseScore/wiki/Set-up-developer-environment and the other pages mentioned in the side bar there under "Compilation".

**Reply by iiicakeeiii** [reply]
Thanks so much for the guidance and resources. Hope I get to a level where I see myself helping out in the future ✌️

**Reply by cbjeukendrup** [reply]
Great! And if you have any questions or when you're experiencing build problems for example, feel free to reach out to us. The easiest way is by using the [MuseScore Developers Discord Server](https://discord.gg/HwHhXEbJ4r).
